qiskit/aqua/operators/state_fns/state_fn.py

```diff
@@ -0,0 +1,353 @@
+
+    def __repr__(self) -> str:
+        return "{}({}, coeff={}, is_measurement={})".format(self.__class__.__name__,
+                                                            repr(self.primitive),
+                                                            self.coeff, self.is_measurement)
+
+    def eval(self,
+             front: Union[str, dict, np.ndarray,
+                          OperatorBase] = None) -> Union[OperatorBase, float, complex]:
+        raise NotImplementedError
+
+    def bind_parameters(self, param_dict: dict) -> OperatorBase:
+        param_value = self.coeff
+        if isinstance(self.coeff, ParameterExpression):
+            unrolled_dict = self._unroll_param_dict(param_dict)
+            if isinstance(unrolled_dict, list):
+                # pylint: disable=import-outside-toplevel
+                from ..list_ops.list_op import ListOp
+                return ListOp([self.bind_parameters(param_dict) for param_dict in unrolled_dict])
+            coeff_param = list(self.coeff.parameters)[0]
+            if coeff_param in unrolled_dict:
+                # TODO what do we do about complex?
+                value = unrolled_dict[coeff_param]
+                param_value = float(self.coeff.bind({coeff_param: value}))
+        return self.__class__(self.primitive, is_measurement=self.is_measurement, coeff=param_value)
+
+    # Try collapsing primitives where possible. Nothing to collapse here.
+    def reduce(self) -> OperatorBase:
+        return self
+
+    def traverse(self,
+                 convert_fn: Callable,
+                 coeff: Optional[Union[int, float, complex, ParameterExpression]] = None
+                 ) -> OperatorBase:
+        r"""
+        Apply the convert_fn to the internal primitive if the primitive is an Operator (as in
+        the case of ``OperatorStateFn``). Otherwise do nothing. Used by converters.
+
+        Args:
+            convert_fn: The function to apply to the internal OperatorBase.
+            coeff: A coefficient to multiply by after applying convert_fn.
+
+        Returns:
+            The converted StateFn.
+        """
+        if isinstance(self.primitive, OperatorBase):
+            return StateFn(convert_fn(self.primitive),
+                           coeff=coeff or self.coeff, is_measurement=self.is_measurement)
+        else:
+            return self
+
+    def to_matrix_op(self, massive: bool = False) -> OperatorBase:
+        """ Return a ``VectorStateFn`` for this ``StateFn``.
+
+        Args:
+            massive: Whether to allow large conversions, e.g. creating a matrix representing
+                over 16 qubits.
+
+        Returns:
+            A VectorStateFn equivalent to self.
+        """
+        # pylint: disable=cyclic-import,import-outside-toplevel
+        from .vector_state_fn import VectorStateFn
+        return VectorStateFn(self.to_matrix(massive=massive), is_measurement=self.is_measurement)
+
+    # TODO to_dict_op
```

Comment:
I guess we don't need this at present then?

Reply:
It's a future TODO, we should add it